File: src/domain/harness.ts
import { z } from "zod";

export type HarnessName = "claude-code" | "codex" | "cursor" | "copilot-cli";

export const CHEESE_DIR = ".cheese";

export const PORTABLE_EVENTS = [
  "sessionStart",
  "preToolUse",
  "postToolUse",
] as const;
export type PortableEvent = (typeof PORTABLE_EVENTS)[number];

export type HookEntry = {
  type: string;
  command: string;
  timeout?: number;
};

const hookEntrySchema = z.object({
  type: z.string().min(1),
  command: z.string().min(1),
  timeout: z.number().int().positive().optional(),
});

export const hooksSourceSchema = z.record(z.string(), z.array(hookEntrySchema));
export type HooksSource = Partial<Record<string, HookEntry[]>>;
export type PortableHooks = Partial<Record<PortableEvent, HookEntry[]>>;

export const pluginMetadataSchema = z.object({
  name: z.string().min(1),
  version: z.string().min(1),
  description: z.string().min(1),
  author: z.object({
    name: z.string().min(1),
    email: z.string().optional(),
    url: z.string().optional(),
  }),
  license: z.string().min(1),
  repository: z.string().min(1),
  homepage: z.string().optional(),
  keywords: z.array(z.string()).optional(),
});

export type PluginMetadata = z.infer<typeof pluginMetadataSchema>;

export type McpServerConfig = {
  command: string;
  args: string[];
  env?: Record<string, string>;
};

export const canonicalMcpServers: Record<string, McpServerConfig> = {
  milknado: {
    command: "npx",
    args: ["tsx", "src/index.ts", "mcp"],
  },
  tilth: { command: "npx", args: ["tilth", "--mcp", "--edit"] },
  context7: {
    command: "npx",
    args: ["-y", "@upstash/context7-mcp"],
  },
  tavily: {
    command: "npx",
    args: ["-y", "tavily-mcp@latest"],
    // biome-ignore lint/suspicious/noTemplateCurlyInString: MCP runtime performs env var substitution on the literal "${TAVILY_API_KEY}" pattern.
    env: { TAVILY_API_KEY: "${TAVILY_API_KEY}" },
  },
};

export type SurfaceEmissionResult = {
  rules: string[];
  commands: string[];
};

export type ManifestComponentPaths = Partial<{
  agents: string;
  skills: string;
  commands: string;
  hooks: string;
  mcpServers: string;
  rules: string;
  apps: string;
}>;

export type AgentFrontmatterFields = {
  name: string;
  description: string;
  tools: string[];
  skills: string[];
  color?: string | undefined;
  effort?: string | undefined;
  disallowedTools: string[];
  permissionMode?: string | undefined;
};

export type AgentArtifactInput = {
  frontmatter: AgentFrontmatterFields;
  resolvedModel: string;
};

export type AgentArtifact = {
  frontmatter: Record<string, unknown>;
  appendix: string;
};

export type HarnessCapabilities = {
  skillFrontmatterKeys: ReadonlySet<string>;
  agentFrontmatterKeys: ReadonlySet<string>;
  hookEvents: ReadonlySet<string>;
  toolNames: ReadonlySet<string>;
  bootstrapHook: boolean;
};

export interface HarnessAdapter {
  name: HarnessName;
  displayName: string;
  outputRoot: string;
  agentDirectory: string;
  skillDirectory: string;
  commandDirectory?: string;
  defaultModel: string;
  notes: string[];

  manifestDir: string;
  buildManifest(
    metadata: PluginMetadata,
    componentPaths: ManifestComponentPaths,
  ): Record<string, unknown>;

  mcpFileName: string;

  buildHookConfig(portable: PortableHooks): Record<string, unknown> | null;

  buildAgentArtifact(input: AgentArtifactInput): AgentArtifact;

  emitSurface?: (
    skillsDir: string,
    outputRoot: string,
  ) => Promise<SurfaceEmissionResult>;

  capabilities: HarnessCapabilities;
}

File: python/cheese_flow/lib/harness.py
"""Port of `src/domain/harness.ts` — harness domain types.

Per the migration spec's Implementation Notes, the TS domain types live on
the slice as `python/cheese_flow/lib/harness.py`. This module is the typed
seam that adapters and (later) the compile pipeline both consume.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import (
    Any,
    Literal,
    TypedDict,
)

CHEESE_DIR = ".cheese"

HarnessName = Literal["claude-code", "codex", "cursor", "copilot-cli"]

PortableEvent = Literal["sessionStart", "preToolUse", "postToolUse"]
PORTABLE_EVENTS: tuple[PortableEvent, ...] = (
    "sessionStart",
    "preToolUse",
    "postToolUse",
)


class _HookEntryRequired(TypedDict):
    type: str
    command: str


class HookEntry(_HookEntryRequired, total=False):
    """Mirror of `HookEntry` (`type`, `command` required; `timeout` optional)."""

    timeout: int


HooksSource = dict[str, list[HookEntry]]
PortableHooks = dict[PortableEvent, list[HookEntry]]


class _PluginAuthorRequired(TypedDict):
    name: str


class PluginAuthor(_PluginAuthorRequired, total=False):
    email: str
    url: str


class _PluginMetadataRequired(TypedDict):
    name: str
    version: str
    description: str
    author: PluginAuthor
    license: str
    repository: str


class PluginMetadata(_PluginMetadataRequired, total=False):
    """Mirror of `PluginMetadata` from `pluginMetadataSchema`."""

    homepage: str
    keywords: list[str]


class _McpServerConfigRequired(TypedDict):
    command: str
    args: list[str]


class McpServerConfig(_McpServerConfigRequired, total=False):
    env: dict[str, str]


CANONICAL_MCP_SERVERS: dict[str, McpServerConfig] = {
    "milknado": {
        "command": "npx",
        "args": ["tsx", "src/index.ts", "mcp"],
    },
    "tilth": {"command": "npx", "args": ["tilth", "--mcp", "--edit"]},
    "context7": {
        "command": "npx",
        "args": ["-y", "@upstash/context7-mcp"],
    },
    "tavily": {
        "command": "npx",
        "args": ["-y", "tavily-mcp@latest"],
        "env": {"TAVILY_API_KEY": "${TAVILY_API_KEY}"},
    },
}


class SurfaceEmissionResult(TypedDict):
    rules: list[str]
    commands: list[str]


class ManifestComponentPaths(TypedDict, total=False):
    agents: str
    skills: str
    commands: str
    hooks: str
    mcpServers: str
    rules: str
    apps: str


class _AgentFrontmatterRequired(TypedDict):
    name: str
    description: str
    tools: list[str]
    skills: list[str]
    disallowedTools: list[str]


class AgentFrontmatterFields(_AgentFrontmatterRequired, total=False):
    """Narrow record passed to `build_base_agent_artifact`.

    Required keys: `name`, `description`, `tools`, `skills`, `disallowedTools`.
    Optional keys: `color`, `effort`, `permissionMode`.
    """

    color: str
    effort: str
    permissionMode: str


@dataclass(frozen=True)
class AgentArtifactInput:
    frontmatter: AgentFrontmatterFields
    resolvedModel: str


@dataclass(frozen=True)
class AgentArtifact:
    frontmatter: dict[str, Any]
    appendix: str


@dataclass(frozen=True)
class HarnessCapabilities:
    skillFrontmatterKeys: frozenset[str]
    agentFrontmatterKeys: frozenset[str]
    hookEvents: frozenset[str]
    toolNames: frozenset[str]
    bootstrapHook: bool


BuildManifest = Callable[[PluginMetadata, ManifestComponentPaths], dict[str, Any]]
BuildHookConfig = Callable[[PortableHooks], dict[str, Any] | None]
BuildAgentArtifact = Callable[[AgentArtifactInput], AgentArtifact]
EmitSurface = Callable[[str, str], Awaitable[SurfaceEmissionResult]]


@dataclass(frozen=True)
class HarnessAdapter:
    """Mirror of the TS `HarnessAdapter` interface as a frozen dataclass."""

    name: HarnessName
    displayName: str
    outputRoot: str
    agentDirectory: str
    skillDirectory: str
    defaultModel: str
    notes: tuple[str, ...]
    manifestDir: str
    mcpFileName: str
    buildManifest: BuildManifest
    buildHookConfig: BuildHookConfig
    buildAgentArtifact: BuildAgentArtifact
    capabilities: HarnessCapabilities
    commandDirectory: str | None = None
    emitSurface: EmitSurface | None = None
